File: Coursica/LoginViewController.m
//
//  LoginViewController.m
//  Coursica
//
//  Created by Matthew Beatty on 1/21/15.
//  Copyright (c) 2015 Prestige Worldwide. All rights reserved.
//

#import "LoginViewController.h"
#import "AppDelegate.h"
#import "CoursesViewController.h"
#import <QuartzCore/QuartzCore.h>

#define coursicaBlue [UIColor colorWithRed:31/255.0 green:148/255.0 blue:255/255.0 alpha:1.0]

@interface LoginViewController () <UIWebViewDelegate, UITextFieldDelegate>

@property (weak, nonatomic) IBOutlet UITextField *usernameField;
@property (weak, nonatomic) IBOutlet UITextField *passwordField;
@property (weak, nonatomic) IBOutlet UIButton *loginButton;
@property (weak, nonatomic) IBOutlet UIWebView *secretWebView;
@property (weak, nonatomic) IBOutlet UILabel *messageLabel;

@property (weak, nonatomic) IBOutlet NSLayoutConstraint *messageHUIDSpace;
@property (weak, nonatomic) IBOutlet NSLayoutConstraint *titlePasswordSpace;
@property (weak, nonatomic) IBOutlet UIView *keyboardView;
@property (weak, nonatomic) IBOutlet NSLayoutConstraint *keyboardHeight;

@property (weak, nonatomic) IBOutlet UITableView *tableView;
@property (assign) BOOL PINSiteLoaded;
@property (assign) BOOL userDidSubmit;
@property (assign) BOOL PINSiteTried;
@property (nonatomic, assign) CGFloat titleTopSpaceInitial;
@property (nonatomic, assign) CGFloat titleTopSpaceError;
@property (assign) BOOL UIStateError;

@end

@implementation LoginViewController

/* This works by loading a UIWebView behind the main login view. When the login view loads for the first time,
 we make a request to CS50 courses for its authentication. 
 When we load this authentication page, self.PINSiteLoaded is set to YES
 When the user submits their credentials, self.userDidSubmit is set to YES
 If these are both set to YES, we try their credentials, and if the authentication website routes us back to cs50 courses,
 we know that login was successful and can dismiss the login controller
 */

- (void)viewDidLoad {
    
    [super viewDidLoad];
    
    self.PINSiteLoaded = NO;
    self.userDidSubmit = NO;
    self.PINSiteTried = NO;
    self.UIStateError = NO;
    NSURLRequest *request = [NSURLRequest requestWithURL:[NSURL URLWithString:@"https://courses.cs50.net/classes/login"]];
    [self.secretWebView loadRequest:request];
    
    self.usernameField.layer.cornerRadius = 4.0f;
    self.passwordField.layer.cornerRadius = 4.0f;
    
    self.usernameField.layer.sublayerTransform = CATransform3DMakeTranslation(10, 0, 0);
    self.passwordField.layer.sublayerTransform = CATransform3DMakeTranslation(10, 0, 0);
    
    self.loginButton.layer.cornerRadius = 2;
    self.loginButton.clipsToBounds = YES;
    
    self.titleTopSpaceInitial = self.titlePasswordSpace.constant;
    
    CGFloat height = [UIScreen mainScreen].bounds.size.height;
    if (height < 568) {
        self.messageHUIDSpace.constant -= 8;
    }
    
    [self registerKeyboardNotifications];
    
    UIImageView *backgroundImageView =
        [[UIImageView alloc] initWithImage:[UIImage imageNamed:@"jharvard_dark.jpg"]];
    backgroundImageView.frame = self.view.bounds;
    [self.view addSubview:backgroundImageView];
    [self.view sendSubviewToBack: backgroundImageView];
    
    [self.view setNeedsLayout];
    [self setNeedsStatusBarAppearanceUpdate];
}

- (CGFloat)titleTopSpaceError {
    
    // TODO: Handle the error message animation on the 4s in a less sketchy way
    
    CGFloat shift = 25;
    
    CGFloat height = [UIScreen mainScreen].bounds.size.height;
    if (height < 568) {
        shift = 5;
    }
    
    return _titleTopSpaceInitial + shift;
}

- (void)tryUserCredentials {
    
    [self removeFailureUIState];
    
    [self.secretWebView stringByEvaluatingJavaScriptFromString:@"setContent('PIN','Harvard University ID (HUID)');"];
    
    NSString *username = self.usernameField.text;
    NSString *usernameJS  = [NSString stringWithFormat:@"document.getElementById('username').value = '%@'", username];
    [self.secretWebView stringByEvaluatingJavaScriptFromString:usernameJS];
    
    NSString *password = self.passwordField.text;
    NSString *passwordJS  = [NSString stringWithFormat:@"document.getElementById('password').value = '%@'", password];
    [self.secretWebView stringByEvaluatingJavaScriptFromString:passwordJS];
    
    NSString *submitJS = @"document.getElementsByName('_eventId_submit')[0].click()";
    self.PINSiteTried = YES;
    [self.secretWebView stringByEvaluatingJavaScriptFromString:submitJS];
}

- (void)webView:(UIWebView *)webView didFailLoadWithError:(NSError *)error {
    
    [self loginFailedWithMessage:@"Error connecting to the network."];
}

- (BOOL)webView:(UIWebView *)webView shouldStartLoadWithRequest:(NSURLRequest *)request navigationType:(UIWebViewNavigationType)navigationType {
    
    if ([[webView.request.URL absoluteString] isEqualToString:@"https://courses.cs50.net/"]) {
        
        [self.usernameField resignFirstResponder];
        [self.passwordField resignFirstResponder];
        
        [self.delegate userDidLogin];
    }
    return YES;
}

- (void)webViewDidFinishLoad:(UIWebView *)webView {
    
    self.PINSiteLoaded = YES;
    if (self.userDidSubmit && !self.PINSiteTried) {
        [self tryUserCredentials];
    } else if (self.PINSiteTried) {
        [self loginFailedWithMessage:@"Invalid HUID or password, try again."];
    }
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

-(UIStatusBarStyle)preferredStatusBarStyle{
    return UIStatusBarStyleLightContent;
}

- (void)removeFailureUIState {
    
    if (!self.UIStateError) {
        return;
    }
    
    if (self.titleTopSpaceError == self.titleTopSpaceError)
        self.titlePasswordSpace.constant = self.titleTopSpaceInitial;

    [UIView animateWithDuration:0.2 animations:^{
        self.messageLabel.alpha = 0.0;
        [self.view layoutIfNeeded];
        // Change to red failure color
        self.loginButton.backgroundColor = coursicaBlue;
    } completion:^(BOOL finished) {
        self.UIStateError = !finished;
    }];
}

- (void)loginFailedWithMessage:(NSString*)message {
    
    self.UIStateError = YES;
    self.messageLabel.text = message;
    self.titlePasswordSpace.constant = self.titleTopSpaceError;
        
    [self.loginButton setTitle:@"Log in" forState:UIControlStateNormal];
    [UIView animateWithDuration:0.2 animations:^{
        self.messageLabel.alpha = 1.0;
        [self.view layoutIfNeeded];
        // Change to red failure color
        self.loginButton.backgroundColor = [UIColor colorWithRed:1.0 green:30/255.0 blue:31/255.0 alpha:1.0];
    }];
}

- (IBAction)loginButtonPressed:(id)sender {

    if (self.usernameField.text.length == 0) {
        [self loginFailedWithMessage:@"Your HUID is required to log in."];
        return;
    }
    if (self.passwordField.text.length == 0) {
        [self loginFailedWithMessage:@"Your password is required to log in."];
        return;
    }
    
    self.userDidSubmit = YES;
    [self.loginButton setTitle:@"Logging in..." forState:UIControlStateNormal];
    if (self.PINSiteLoaded) {
        [self tryUserCredentials];
    }
}

- (void)registerKeyboardNotifications {
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(keyboardChangeFrame:)
                                                 name:UIKeyboardWillShowNotification
                                               object:nil];
    
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(keyboardChangeFrame:)
                                                 name:UIKeyboardWillHideNotification
                                               object:nil];
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(keyboardChangeFrame:)
                                                 name:UIKeyboardWillChangeFrameNotification
                                               object:nil];
}

-(void)keyboardChangeFrame:(NSNotification*)notification{
    
    CGRect keyboardEndFrame;
    [[notification userInfo][UIKeyboardFrameEndUserInfoKey] getValue:&keyboardEndFrame];
    UIViewAnimationCurve curve =[notification.userInfo[UIKeyboardAnimationCurveUserInfoKey] integerValue];
    NSTimeInterval duration = [notification.userInfo[UIKeyboardAnimationDurationUserInfoKey] doubleValue];
    self.keyboardHeight.constant = CGRectGetHeight(keyboardEndFrame);
    [self.view setNeedsUpdateConstraints];
    [self.view setNeedsLayout];
    
    [UIView beginAnimations:nil context:NULL];
    [UIView setAnimationDuration:duration];
    [UIView setAnimationCurve:curve];
    [UIView setAnimationBeginsFromCurrentState:YES];
    
    [self.view layoutIfNeeded];
    
    [UIView commitAnimations];
}

@end

File: Coursica/LoginViewController.swift
//
//  LoginViewController.swift
//  Coursica
//
//  Created by Regan Bell on 8/2/15.
//  Copyright (c) 2015 Prestige Worldwide. All rights reserved.
//

import UIKit

protocol LoginViewControllerDelegate {
    
    func userDidLoginWithHUID(HUID: String)
}

class LoginViewController: CoursicaViewController {

    @IBOutlet var usernameField: UITextField!
    @IBOutlet var passwordField: UITextField!
    @IBOutlet var loginButton: UIButton!
    @IBOutlet var errorMessageLabel: UILabel!
    
    @IBOutlet var messageHUIDSpace: NSLayoutConstraint!
    @IBOutlet var titlePasswordSpace: NSLayoutConstraint!
    @IBOutlet var keyboardHeight: NSLayoutConstraint!
    
    @IBOutlet var keyboardView: UIView!
    var delegate: LoginViewControllerDelegate!
    
    var userDidSubmit = false
    var PINSiteTried = false
    var UIStateError = false
    var allowAnyLogin = false
    var screenIsSmall: Bool { get { return UIScreen.mainScreen().bounds.size.height < 568 } }
    var titleTopSpaceInitial: CGFloat = 0
    var titleTopSpaceError: CGFloat { get { return self.titleTopSpaceInitial + (self.screenIsSmall ? 5 : 25) } }
    var secondsWaitedByUser = 0
    
    func checkAllowAnyLogin() {
        
        let urlString = "glaring-heat-9505.firebaseIO.com/allowAnyLogin"
        let root = Firebase(url: urlString)
        weak var weakSelf = self
        root.observeSingleEventOfType(FEventType.Value, withBlock: {snapshot in
            if let flag = snapshot.value as? Bool {
                weakSelf?.allowAnyLogin = flag
            }
        })
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        checkAllowAnyLogin()

        usernameField.layer.cornerRadius = 4.0
        passwordField.layer.cornerRadius = 4.0
        
        usernameField.layer.sublayerTransform = CATransform3DMakeTranslation(10, 0, 0)
        passwordField.layer.sublayerTransform = CATransform3DMakeTranslation(10, 0, 0)
        
        usernameField.returnKeyType = .Next
        passwordField.returnKeyType = .Go
        
        usernameField.delegate = self
        passwordField.delegate = self
        
        loginButton.layer.cornerRadius = 2
        loginButton.clipsToBounds = true
        
        titleTopSpaceInitial = titlePasswordSpace.constant;
        
        if (UIScreen.mainScreen().bounds.size.height < 568) {
            messageHUIDSpace.constant -= 8;
        }
        
        self.registerKeyboardNotifications()
        
        let backgroundImageView = UIImageView(image: UIImage(named: "jharvard_dark.jpg"))
        backgroundImageView.frame = self.view.bounds
        self.view.addSubview(backgroundImageView)
        self.view.sendSubviewToBack(backgroundImageView)
    }
    
    func waitingTitle() -> String {
        switch secondsWaitedByUser {
        case 0, 1: return "Logging in..."
        case 2, 3, 4: return "Still working..."
        default: return "Almost there..."
        }
    }
    
    func updateWaitingMessage(timer: NSTimer) {
        
        secondsWaitedByUser++
        if let _ = errorMessageLabel.text {
            timer.invalidate()
            return
        }
        loginButton.setTitle(self.waitingTitle(), forState: .Normal)
    }
    
    func displayErrorMessage(errorMessage: String?) {
        let errorState = errorMessage != nil
        errorMessageLabel.text = errorMessage
        titlePasswordSpace.constant = errorState ? titleTopSpaceError : titleTopSpaceInitial
        loginButton.setTitle("Log in", forState: .Normal)
        UIView.animateWithDuration(0.3, animations: {
            self.errorMessageLabel.alpha = errorState ? 1 : 0
            self.view.layoutIfNeeded()
            self.loginButton.backgroundColor = errorState ? UIColor(red: 1, green: 30/255.0, blue: 31/255.0, alpha: 1) : coursicaBlue
        })
    }
    
    @IBAction func loginButtonPressed(button: UIButton?) {
        guard let username = usernameField.text where !username.isEmpty else {
            displayErrorMessage("Your HUID is required to log in.")
            return
        }
        guard let password = passwordField.text where !password.isEmpty else {
            displayErrorMessage("Your password is required to log in.")
            return
        }
        loginButton.setTitle("Logging in...", forState: .Normal)
        secondsWaitedByUser = 0
        NSTimer.scheduledTimerWithTimeInterval(1, target: self, selector: "updateWaitingMessage:", userInfo: nil, repeats: true)
        weak var weakSelf: LoginViewController! = self
        Login.attemptLoginWithCredentials(username, password: password, completionBlock: {success, error in
            if success || weakSelf.allowAnyLogin {
                weakSelf.usernameField.resignFirstResponder()
                weakSelf.passwordField.resignFirstResponder()
                weakSelf.delegate.userDidLoginWithHUID(username)
                weakSelf.dismissViewControllerAnimated(true, completion: nil)
            } else {
                weakSelf.displayErrorMessage(error)
            }
        })
    }
    
    // MARK: Keyboard
    
    func registerKeyboardNotifications() {
        NSNotificationCenter.defaultCenter().addObserver(self, selector: "keyboardChangeFrame:", name: UIKeyboardWillShowNotification, object: nil)
        NSNotificationCenter.defaultCenter().addObserver(self, selector: "keyboardChangeFrame:", name: UIKeyboardWillHideNotification, object: nil)
        NSNotificationCenter.defaultCenter().addObserver(self, selector: "keyboardChangeFrame:", name: UIKeyboardWillChangeFrameNotification, object: nil)
    }
    
    func keyboardChangeFrame(notification: NSNotification) {
        
        let endFrame = (notification.userInfo![UIKeyboardFrameEndUserInfoKey] as! NSValue).CGRectValue()
        let curve = (notification.userInfo![UIKeyboardAnimationCurveUserInfoKey] as! NSNumber).integerValue
        let duration = (notification.userInfo![UIKeyboardAnimationDurationUserInfoKey] as! NSNumber).doubleValue
        keyboardHeight.constant = endFrame.height
        
        UIView.beginAnimations(nil, context: nil)
        UIView.setAnimationDuration(duration as CFTimeInterval)
        UIView.setAnimationCurve(UIViewAnimationCurve(rawValue: curve)!)
        UIView.setAnimationBeginsFromCurrentState(true)
        self.view.layoutIfNeeded()
        UIView.commitAnimations()
    }
}

extension LoginViewController: UITextFieldDelegate {
    
    func textFieldShouldReturn(textField: UITextField) -> Bool {
        if textField.returnKeyType == .Next {
            passwordField.becomeFirstResponder()
        } else {
            self.loginButtonPressed(nil)
        }
        return true
    }
}

